### ISSUE TITLE
Changed behaviour between 0.2 and 0.3 leads to data exposure if undefined value provided

### ISSUE THREAD

**Issue by gugu** [typeorm/typeorm#9027]
## Issue Description

Version 0.2 treats undefined value as null while version 0.3 omits undefined values from the query. This changes code behaviour drastically, for example (express.js):

```typescript
UserRepository.find({ where: { email: req.body.email, password: req.body.password } })
// and attacker sends req.body
{}
```

```mysql
-- 0.2
SELECT * FROM Users WHERE email=null AND password = null -- nothing found, authorization fails
-- 0.3
SELECT * FROM Users -- returns user, authorization process passes
```
and
```typescript
PasswordRecoveryToke.find({ where: { token: req.query.token } })
// and attacker removes token parameter from query string, so req.query = {}
```

```mysql
-- 0.2
SELECT * FROM PasswordRecoveryToken WHERE token = null -- nothing found, password recovery fails
-- 0.3
SELECT * FROM PasswordRecoveryToken -- token found, password successfully changed
```

### Expected Behavior

Values coming from third-party should not change SQL query plan. There are three (in my opinion) correct ways to handle undefined:
1. 0.2 way - treat undefined as null
2. Exception way - throw an exception if TypeORM receives undefined and introduce OmitValue() operator similar to isNull() if we really want to omit a column
3. Empty query way - use `0 = 1` if undefined passed to make query return 0 values  and introduce OmitValue() operator similar to isNull() if we really want to omit a column

### Actual Behavior

Selecting every row in the table

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | yes            |
| `aurora-postgres` | yes            |
| `better-sqlite3`  | ?            |
| `cockroachdb`     | ?            |
| `cordova`         | ?            |
| `expo`            | ?            |
| `mongodb`         | ?            |
| `mysql`           | yes            |
| `nativescript`    | ?            |
| `oracle`          | ?            |
| `postgres`        | ?            |
| `react-native`    | ?            |
| `sap`             | ?            |
| `spanner`         | ?            |
| `sqlite`          | ?            |
| `sqlite-abstract` | ?            |
| `sqljs`           | ?            |
| `sqlserver`       | ?            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✅ Yes, I have the time, and I know how to start.

**Comment by mattrunyon** [comment]
This doesn't seem like a bug. You should be sanitizing any user provided inputs to prevent this.

From the view of typeorm, an undefined where parameter should not create a where clause.

**Comment by gugu** [comment]
> This doesn't seem like a bug. You should be sanitizing any user provided inputs to prevent this.
> 
> From the view of typeorm, an undefined where parameter should not create a where clause.

Changing behavior of query based on user's input leading to exposure of data is SQL injection. It is very easy to get undefined value in web applications, and one under-validation exposes whole business to malicious actors. The decision is okay for ORM itself, but it is extremely dangerous in most of ORM applications

**Comment by draaglom** [comment]
big ➕  here. Good API design is safe by default. Expecting people to validate when the path of least resistance is to not do that will inevitably lead to a world of pain for library consumers. 

By analogy:

"This doesn't seem like a bug. You should remember to free() after you malloc()"
"This doesn't seem like a bug. You should be bounds-checking your arrays"

^ two statements which are nice in theory but led to the invention of a whole new programming language ecosystem (C -> Rust) to get around them ;)

**Comment by pleerock** [comment]
`undefined` values are skipped. This is by design.

Why not to allow passing values the following way:

```ts
UserRepository.find({ where: req.body.where })
```

```ts
UserRepository.find({ where: req.body.where })
```

```ts
UserRepository.find(req.body)
```

We do sanitizing anyway right? So, it's exactly like @mattrunyon said: "From the view of typeorm, an undefined where parameter should not create a where clause".

One can ask why `undefined` leads to `null` and he will be right - why? `undefined` and `null` are different things and should not be confused.

If we want to blame some design, then I better to blame JavaScript's `undefined` and `null`.
We all know that problems they bring into JavaScript, but we always forget that they also brings us many opportunities.
Some great JavaScript features that we were never think about exist because of `undefined`.

User face different cases.
Solution in your case can be simple as:

```ts
UserRepository.find({ where: { 
   email: req.body.email || null, 
   password: req.body.password || null 
}})
```

Others might benefit from skipping `undefined` values in case of complex requests, for example searching by either `firstName` or `lastName`:

```ts
UserRepository.find({ where: { 
   firstName: req.body.firstName,
   lastName: req.body.lastName
}})
```

**Comment by namhtpyn** [comment]
undefined should never be convert to null in the first place. It seems like they were doing it wrong in 0.2 and have fixed that in 0.3. 👍 for that.

**Comment by scarfunk** [comment]
**DONT**
```
UserRepository.find({ where: { 
   email: req.body.email || null, 
   password: req.body.password || null 
}})
```
When the value is false, it is converted to null, so it is recommended to use the following syntax

**DO**
```
UserRepository.find({ where: { 
   email: req.body.email ?? null, 
   password: req.body.password ?? null 
}})
```
